### ISSUE TITLE
Auto-disconnect

### ISSUE THREAD

**Issue by schickling** [prisma/prisma-client-js#540]
Right now users need to explicitly call `await prisma.disconnect()` at the end of their program otherwise the running Node process won't terminate.

Ideally the Prisma Client (or more correctly: the used Prisma Query Engine) should automatically disconnect when the Node process is trying to terminate. This removes a lot of friction points and simplifies the usage of Prisma Client - especially when getting started - since users don't need to call `await prisma.disconnect()` explicitly anymore.

**Comment by sorenbs** [comment]
NEXT STEP: @schickling and @timsuchanek have learned about a potential approach during their trip to SF. They will write it up here. Then later @sorenbs will write up a proposal based on that and other considerations.

**Comment by timsuchanek** [comment]
The assumption was, that if we start the Prisma process as a `detached` child process, Node.js still could properly exit, once the event loop is empty.

However, the problem is, that we still need to keep a stream to the process' stderr and stdout open. These streams will still prevent the Node.js process from exiting.

We can revisit this topic once we have replaced the HTTP server with Neon bindings, as that will fundamentally change the interaction between Node.js and Rust for Prisma Client.

**Comment by jedmao** [comment]
I had a full page of ideas that I deleted upon this final conclusion. I hope you find it valuable.

```ts
class PrismaClient {
    public async useOnce(callback: (this: this) => Promise<void>) {
        try {
            await callback.call(this);
        } finally {
            this.disconnect();
        }
    }
}

new PrismaClient().useOnce(async function() {
    const post = await this.post.update({
        where: { id: 1 },
        data: { published: true }
    });
    console.log(post);
});
```

One caveat is if someone didn't know JavaScript very well and attempted to provide a lambda function, but they would immediately realize their mistake when there was no type information after typing `this.`.

**Comment by timsuchanek** [comment]
Thanks @jedmao! That's an interesting API proposal.
And of course using proper `private` properties :)
That would indeed solve these cases in which you just want to do a query and that's it.

**Comment by adarnon** [comment]
@jedmao I think the `return this.disconnect()` call should have `await`?

**Comment by jedmao** [comment]
> @jedmao I think the `return this.disconnect()` call should have `await`?

Nope. An async function returns a promise. Disconnect already returns a promise, so I’m intentionally returning that one, rather than wrapping it in yet another one.
